snooty/parser.py

```diff
@@ -1702,10 +1700,45 @@ def postprocess(self) -> PostprocessorResult:
 
         return result
 
-    def _page_updated(self, page: Page, diagnostics: List[Diagnostic]) -> None:
+    def parse_rst_files(
+        self, paths: Iterable[Path], max_workers: Optional[int] = None
+    ) -> None:
+        pool = multiprocessing.Pool(max_workers)
+        try:
+            logger.debug("Processing rst files")
+            cache_misses: List[Path] = []
+
+            hits = 0
+            for path in paths:
+                try:
+                    page, diagnostics = self.cache.get(self.config, path)
+                    self._page_updated(page, diagnostics)
+                    hits += 1
+                except KeyError:
+                    cache_misses.append(path)
+
+            logger.info("cache: %d hits and %d misses", hits, len(cache_misses))
+
+            results = pool.imap_unordered(partial(parse_rst, self.parser), cache_misses)
+            for sequence in results:
+                for page, diagnostics in sequence:
+                    self._page_updated(page, diagnostics)
+
+        finally:
+            # We cannot use the multiprocessing.Pool context manager API due to the following:
+            # https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html#if-you-use-multiprocessing-pool
+            pool.close()
+            pool.join()
+
+    def update_cache(self, optimize: bool = False) -> None:
+        cache = parse_cache.CacheData(self.cache_file.generate_specifier(), {})
+        self.pages.add_to_cache(cache)
+        self.cache_file.persist(cache, optimize=optimize)
+
+    def _page_updated(self, page: Page, diagnostics: Sequence[Diagnostic]) -> None:
         """Update any state associated with a parsed page."""
         # Finish any pending tasks
-        page.finish(diagnostics, self)
+        page.finish(list(diagnostics), self)
```

Comment:
(Nit) We save another instance of `list(diagnostics)` as a variable `diagnostics_copy` on line 1781. Should we just reuse the variable here to avoid making a new list? Or is there a reason we make a separate list copy?

Reply:
honestly some type tweaks would be nice so we don't need to create a list copy at all, but done!